### PR TITLE
Add Intel CET to getTargetInfo

### PR DESCRIPTION
**Pull request by ghost**
This follows PR #15415 which added Intel CET IBT support and LDC PR [#4437](https://github.com/ldc-developers/ldc/pull/4437) to add support for the new CET target in order to maintain a common interface between DMD and LDC.

Later it would be useful to do the same for GDC as well.

### CODE REVIEW

**Review by ghost — APPROVED** [review]
I suggest to use "intelCET" instead of just "CET" as the later is not very well known.

**Comment by dkorpel** [comment]
Why does D code need to know about this setting?

**Comment by ErnyTech** [comment]
> Why does D code need to know about this setting?

Mainly for inline ASM

**Review by dkorpel — APPROVED** [review]

**Review by RazvanN7 — COMMENTED** [review]
Please add a changelog entry describing this change so that users get to know about this.

**Comment by ErnyTech** [comment]
> Please add a changelog entry describing this change so that users get to know about this.

Added

File: changelog/dmd.intel-cet-ibt-protection.dd [hunk 1]
```diff
@@ -0,0 +1,22 @@
+Added support for Intel CET (Control-flow Enforcement Technology) IBT (Indirect Branch Tracking) protection
+
+CET is a technology that is useful for preventing an attacker from redirecting a program's control flow,
+specifically IBT prevents an an attacker from causing an indirect branch to go to an unintended place
```
Comment (RazvanN7) [inline changelog/dmd.intel-cet-ibt-protection.dd:4]:
```suggestion
specifically IBT prevents an attacker from causing an indirect branch to go to an unintended place.
```

**Review by RazvanN7 — APPROVED** [review]

File: changelog/dmd.intel-cet-ibt-protection.dd [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+To enable Intel IBT protection in DMD you need to pass the `-fIBT` flag to the compiler, consequently the compiler
+will manage the emission of instructions for IBT by itself.
+Be careful when using inline assembly, the compiler will not automatically handle IBT inside an inline assembly.
+
+To find out within a D program whether IBT has been activated or not use the traits getTargetInfo as follows:
+
+---
+// IBT active
+static assert(__traits(getTargetInfo, "CET") == 1);  // CET == 1 if IBT is active
```
Comment (WalterBright) [inline changelog/dmd.intel-cet-ibt-protection.dd:18]: I'm reluctant to call this "CET" when the switch is "IBT". CET includes shadow stack support, which DMD does not yet support. When it does support CET, I'll add a -fCET and only then should getTargetInfo say it supports CET.

**Review by WalterBright — DISMISSED** [review]

File: changelog/dmd.intel-cet-ibt-protection.dd (hunk 2 above)
Comment (ErnyTech) [reply changelog/dmd.intel-cet-ibt-protection.dd:18]: CET includes both IBT and shadow stack, LDC use `__traits(getTargetInfo, "CET") ` as follows:

- Disabled: `__traits(getTargetInfo, "CET") == 0`
- IBT (--fcf-protection=branch in LDC): `__traits(getTargetInfo, "CET") == 1`
- Shadow stack (--fcf-protection=return in LDC): `__traits(getTargetInfo, "CET") == 2`
- IBT + Shadow stack (--fcf-protection=full in LDC): `__traits(getTargetInfo, "CET") == 3`

This behavior is equivalent to `__CET__` of C compilers

Comment (WalterBright) [reply changelog/dmd.intel-cet-ibt-protection.dd:18]: ok then

**Comment by ErnyTech** [comment]
@dkorpel Is there anything blocking this PR? I think it would be good to include it in 2.106.0-beta especially since it also provides the changelog for -fIBT
Walter's request seems satisfied

**Comment by ibuclaw** [comment]
This test is non-portable. Maybe there should be an x86 directory added for such tests (this would include all DMD `asm` tests too).

**Pull request merged**